File: nodejs/src/client/improvements.ts
/**
 * Improvements API: list, apply, reject.
 * Canonical API: GET /ai/improvements, POST /ai/improvements.
 *
 * Requirements: 8.3, 8.4, 8.5, 8.6
 */

import type { LoxtepHttpClient } from '../http/client.js';
import type {
  Improvement,
  ImprovementsListFilters,
  ImprovementsListResponse,
  ImprovementActionResponse,
} from './improvements-types.js';

const IMPROVEMENTS_BASE = '/ai/improvements';

function buildQueryString(params: Record<string, string | number | undefined>): string {
  const search = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined || v === null) continue;
    search.set(k, String(v));
  }
  const qs = search.toString();
  return qs ? `?${qs}` : '';
}

export interface ImprovementsApi {
  /** List improvements for the authenticated organization (R8.3). */
  list: (filters?: ImprovementsListFilters) => Promise<{ improvements: Improvement[]; cursor: string | null }>;
  /** Apply an improvement — sets status to 'applied' on the server (R8.4). */
  apply: (id: string) => Promise<{ id: string; status: 'applied'; updated_at: string }>;
  /** Reject an improvement — sets status to 'rejected' on the server (R8.5). */
  reject: (id: string) => Promise<{ id: string; status: 'rejected'; updated_at: string }>;
}

export function createImprovementsApi(http: LoxtepHttpClient): ImprovementsApi {
  return {
    async list(filters?: ImprovementsListFilters) {
      const params: Record<string, string | number | undefined> = {};
      if (filters?.status) params.status = filters.status;
      if (filters?.workflow_name) params.workflow_name = filters.workflow_name;
      if (filters?.limit != null) params.limit = filters.limit;
      if (filters?.cursor) params.cursor = filters.cursor;
      const qs = buildQueryString(params);
      const res = await http.get<ImprovementsListResponse>(`${IMPROVEMENTS_BASE}${qs}`);
      return res.data;
    },

    async apply(id: string) {
      const res = await http.post<ImprovementActionResponse>(IMPROVEMENTS_BASE, {
        id,
        action: 'apply',
      });
      return res.data as { id: string; status: 'applied'; updated_at: string };
    },

    async reject(id: string) {
      const res = await http.post<ImprovementActionResponse>(IMPROVEMENTS_BASE, {
        id,
        action: 'reject',
      });
      return res.data as { id: string; status: 'rejected'; updated_at: string };
    },
  };
}

File: python/src/loxtep/improvements.py
"""
Improvements API (AI Eval self-improvement). list, apply, reject.
Internal / experimental — not part of the documented customer surface.
Backend: /ai/improvements.
"""

from typing import Any, Optional

from .http_client import AsyncLoxtepHttpClient, LoxtepHttpClient

BASE = "/ai/improvements"


def _unwrap(res: Any) -> Any:
    return res.get("data", res) if isinstance(res, dict) else res


def _query_string(params: dict[str, Any]) -> str:
    parts = [f"{k}={v}" for k, v in params.items() if v is not None]
    return "?" + "&".join(parts) if parts else ""


class ImprovementsApi:
    """Sync improvements surface (internal)."""

    def __init__(self, http: LoxtepHttpClient) -> None:
        self._http = http

    def list(
        self,
        *,
        status: Optional[str] = None,
        workflow_name: Optional[str] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
    ) -> dict[str, Any]:
        qs = _query_string(
            {"status": status, "workflow_name": workflow_name, "limit": limit, "cursor": cursor}
        )
        return _unwrap(self._http.get(f"{BASE}{qs}"))

    def apply(self, id: str) -> dict[str, Any]:
        return _unwrap(self._http.post(BASE, {"id": id, "action": "apply"}))

    def reject(self, id: str) -> dict[str, Any]:
        return _unwrap(self._http.post(BASE, {"id": id, "action": "reject"}))


class AsyncImprovementsApi:
    """Async improvements surface (internal)."""

    def __init__(self, http: AsyncLoxtepHttpClient) -> None:
        self._http = http

    async def list(
        self,
        *,
        status: Optional[str] = None,
        workflow_name: Optional[str] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
    ) -> dict[str, Any]:
        qs = _query_string(
            {"status": status, "workflow_name": workflow_name, "limit": limit, "cursor": cursor}
        )
        return _unwrap(await self._http.get(f"{BASE}{qs}"))

    async def apply(self, id: str) -> dict[str, Any]:
        return _unwrap(await self._http.post(BASE, {"id": id, "action": "apply"}))

    async def reject(self, id: str) -> dict[str, Any]:
        return _unwrap(await self._http.post(BASE, {"id": id, "action": "reject"}))
